src/box/box.cc

```diff
@@ -4449,24 +4476,41 @@ box_process_fetch_snapshot(struct iostream *io,
 		if (replica->gc != NULL)
 			gc_consumer_unregister(replica->gc);
 		replica->gc = gc_consumer_register(
-			instance_vclock, GC_CONSUMER_REPLICA, &replica->uuid);
+			gc_vclock, GC_CONSUMER_REPLICA, &replica->uuid);
 		/* Persist consumer before sending data. */
 		if (gc_consumer_persist(replica->gc) != 0)
 			diag_raise();
+		/* Clear checkpoint reference, if any. */
+		if (replica->gc_checkpoint_ref != NULL) {
+			gc_unref_checkpoint(replica->gc_checkpoint_ref);
+			free(replica->gc_checkpoint_ref);
+			replica->gc_checkpoint_ref = NULL;
+		}
 	}
 
-	say_info("sending read-view to replica at %s", sio_socketname(io->fd));
-	/* Used for checkpoint initial join. */
-	struct checkpoint_cursor cursor;
-	struct checkpoint_cursor *cursor_ptr = NULL;
-	if (req.is_checkpoint_join) {
-		memset(&cursor, 0, sizeof(cursor));
-		cursor.vclock = &req.checkpoint_vclock;
-		cursor.start_lsn = req.checkpoint_lsn;
-		cursor_ptr = &cursor;
+	/* Reference checkpoint in case of checkpoint join. */
+	struct gc_checkpoint_ref checkpoint_ref;
+	auto gc_checkpoint_ref_guard = make_scoped_guard([&]() {
+		gc_unref_checkpoint(&checkpoint_ref);
+	});
+	gc_checkpoint_ref_guard.is_active = false;
+	if (checkpoint != NULL) {
+		if (replica == NULL) {
+			gc_ref_checkpoint(checkpoint, &checkpoint_ref,
+					  "checkpoint join");
+			gc_checkpoint_ref_guard.is_active = true;
+		} else {
+			replica->gc_checkpoint_ref =
+				xalloc_object(struct gc_checkpoint_ref);
```

Comment:
Let's make `gc_ref_checkpoint` allocate an object for us, like `gc_consumer_register` does?

Reply:
Yep, good idea, thanks. Done in a separate commit.